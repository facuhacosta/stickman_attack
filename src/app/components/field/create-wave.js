const generatePosition = (enemies_proximity) => {
  return Math.floor(Math.random() * enemies_proximity) * 5;
};

const generateEnemyType = (enemies) => {
  const randomValue = Math.floor(Math.random() * 100);
  let flag = 0;
  let returnValue = ''
  Object.keys(enemies).some(key => {
    flag += enemies[key];
    if (flag > randomValue) {
      returnValue = key
      return key;
    };
  });
  return (returnValue);
};

const generateDelay = (wave_duration) => (Math.floor(Math.random() * 100) * (wave_duration / 100));

const generateWaveSetting = (wave_number) => {
  let wave_size = 2;
  let wave_duration = 15;
  if (wave_number <= 5) {
    wave_size += wave_number * 2
  } else if (wave_number <= 10) {
    wave_size = 10 + ((wave_number - 5) * 4);
    wave_duration += wave_number + 2
  } else if (wave_number <= 20 ) {
    wave_size = 30 + ((wave_number - 10) * 6);
    wave_duration = 25 + (wave_number * 2)
  } else if (wave_number <= 30) {
    wave_size = 90 + ((wave_number - 10) * 4);
    wave_duration = 45 + (wave_number)
  } else  {
    wave_size = 130 + ((wave_number - 10) * 2);
    wave_duration = 60
  }
  return {wave_size, wave_duration};
}

export const generateWave = (enemies, enemies_proximity, wave_number) => {
  const wavePreset = [];
  const { wave_size, wave_duration } = generateWaveSetting(wave_number)
  for (let index = 0; index < wave_size; index++) {
    const enemyPreset = generateEnemyType(enemies);
    const positionPreset = generatePosition(enemies_proximity);
    const delayPreset = generateDelay(wave_duration);
    wavePreset.push({
      key: index,
      type: `${enemyPreset}`,
      position: `${positionPreset}`,
      delay: delayPreset,
      status: true
    });
  };
  return wavePreset;
}