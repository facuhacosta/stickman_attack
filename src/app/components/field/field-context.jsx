import React, { createContext, useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import gif from '../../../public/assets/man.webm' //make it came from prop or context
import { GameContext } from "../game-context/game-context";
import { generateWave } from './create-wave' 

export const FieldContext = createContext();

export const FieldProvider = (props) => {
  const [points, setPoints] = useState(0);
  const [castleHealth, setCastleHealth] = useState(1000);
  const [gameOver, setGameOver] = useState(false);
  const [enemiesProb, setEnemiesProb] = useState({
    basic: 70,
    powered: 25,
    boss: 5
  });
  let navigate = useNavigate();
  const [enemies, setEnemies] = useState(() => {
    return {
      basic: {
        probability: 70,
        health: 100,
        money_recived: 15,
        damage_dealed: 50,
        runingURL: gif,
        attackingURL: '',
      },
      powered: {
        probability: 25,
        health: 200,
        money_recived: 30,
        damage_dealed: 75,
        runingURL: gif,
        attackingURL: '',
      },
      boss: {
        probability: 5,
        health: 400,
        money_recived: 80,
        damage_dealed: 150,
        runingURL: gif,
        attackingURL: '',
      },
    }
    // fetch('/api/get-enemies/', { body: JSON.stringify({user_wave: props.user.wave})})
  });

  const enemies_proximity = 20;
  const { waveNumber, setWaveNumber } = useContext(GameContext);

  const [wave, setWave] = useState(generateWave(enemiesProb, enemies_proximity, waveNumber));

  useEffect(() => {
    if (castleHealth <= 0 && !gameOver) {
      setGameOver(true);
      setWaveNumber(1);
      castleHealth <= 0 && navigate('/', { replace: true, state: {cameFrom: 'GAME' ,cameWith: 'GAMEOVER'} })
    }
  }, [castleHealth])

  return (
    <FieldContext.Provider value={{
      enemies: enemies,
      points: points,
      setPoints: setPoints,
      castleHealth: castleHealth,
      setCastleHealth: setCastleHealth,
      gameOver: gameOver,
      wave: wave,
      setWave: setWave
    }} >
      {props.children}
    </FieldContext.Provider>
  )
}