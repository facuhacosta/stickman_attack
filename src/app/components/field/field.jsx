import React from 'react';
import style from './field.module.scss';
import { StickmanHandler } from '../stickman-handler';
import { Castle } from '../castle'
import { FieldProvider } from './field-context';
  
export function Field() {

  return (
    <div className={style.field}>
      <FieldProvider>
        <Castle />
        <StickmanHandler />
      </FieldProvider>
    </div>
  );
}