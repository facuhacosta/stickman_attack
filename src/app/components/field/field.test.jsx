import React from 'react';
import { render } from '@testing-library/react';
import { Field } from './field';

describe('Field Component', () => {

  // Matching Snapshots
  test('it should match the snapshot', () => {
    const { asFragment } = render(<Field />);
    expect(asFragment()).toMatchSnapshot();
  });

  // Unit tests

});