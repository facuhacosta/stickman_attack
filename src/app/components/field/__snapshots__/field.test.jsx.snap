// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Component it should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            1000
          </span>
        </div>
      </div>
      <img
        alt="castle"
        src="[object Object]"
      />
    </div>
    <div>
      <span>
        0
      </span>
      <div
        class="undefined"
        style="bottom: 15%; animation-delay: 0s;"
      >
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            200
          </span>
        </div>
        <video
          autoplay=""
          data-testid="stickman-testId"
          loop=""
        >
          <source
            src="[object Object]"
            type="video/webm"
          />
        </video>
      </div>
      <div
        class="undefined"
        style="bottom: 15%; animation-delay: 6s;"
      >
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            100
          </span>
        </div>
        <video
          autoplay=""
          data-testid="stickman-testId"
          loop=""
        >
          <source
            src="[object Object]"
            type="video/webm"
          />
        </video>
      </div>
      <div
        class="undefined"
        style="bottom: 40%; animation-delay: 4s;"
      >
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            200
          </span>
        </div>
        <video
          autoplay=""
          data-testid="stickman-testId"
          loop=""
        >
          <source
            src="[object Object]"
            type="video/webm"
          />
        </video>
      </div>
      <div
        class="undefined"
        style="bottom: 85%; animation-delay: 7s;"
      >
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            100
          </span>
        </div>
        <video
          autoplay=""
          data-testid="stickman-testId"
          loop=""
        >
          <source
            src="[object Object]"
            type="video/webm"
          />
        </video>
      </div>
      <div
        class="undefined"
        style="bottom: 30%; animation-delay: 0s;"
      >
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            200
          </span>
        </div>
        <video
          autoplay=""
          data-testid="stickman-testId"
          loop=""
        >
          <source
            src="[object Object]"
            type="video/webm"
          />
        </video>
      </div>
      <div
        class="undefined"
        style="bottom: 25%; animation-delay: 3s;"
      >
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            400
          </span>
        </div>
        <video
          autoplay=""
          data-testid="stickman-testId"
          loop=""
        >
          <source
            src="[object Object]"
            type="video/webm"
          />
        </video>
      </div>
      <div
        class="undefined"
        style="bottom: 0%; animation-delay: 7s;"
      >
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            400
          </span>
        </div>
        <video
          autoplay=""
          data-testid="stickman-testId"
          loop=""
        >
          <source
            src="[object Object]"
            type="video/webm"
          />
        </video>
      </div>
      <div
        class="undefined"
        style="bottom: 45%; animation-delay: 4s;"
      >
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            100
          </span>
        </div>
        <video
          autoplay=""
          data-testid="stickman-testId"
          loop=""
        >
          <source
            src="[object Object]"
            type="video/webm"
          />
        </video>
      </div>
      <div
        class="undefined"
        style="bottom: 80%; animation-delay: 1s;"
      >
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            400
          </span>
        </div>
        <video
          autoplay=""
          data-testid="stickman-testId"
          loop=""
        >
          <source
            src="[object Object]"
            type="video/webm"
          />
        </video>
      </div>
      <div
        class="undefined"
        style="bottom: 10%; animation-delay: 7s;"
      >
        <div>
          <div
            style="width: 100%;"
          />
          <span>
            200
          </span>
        </div>
        <video
          autoplay=""
          data-testid="stickman-testId"
          loop=""
        >
          <source
            src="[object Object]"
            type="video/webm"
          />
        </video>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
